openSearchLane.go

```diff
@@ -0,0 +1,469 @@
+	offline             bool
+	OpenSearchUrl       string          `json:"openSearchUrl"`
+	OpenSearchPort      string          `json:"openSearchPort"`
+	OpenSearchUser      string          `json:"openSearchUser"`
+	OpenSearchPass      string          `json:"openSearchPass"`
+	OpenSearchIndex     string          `json:"openSearchIndex"`
+	OpenSearchAppName   string          `json:"openSearchAppName"`
+	OpenSearchTransport *http.Transport `json:"openSearchTransport"`
+	LogThreshold        int             `json:"logThreshold,omitempty"`
+	MaxBufferSize       int             `json:"maxBufferSize,omitempty"`
+	BackoffInterval     time.Duration   `json:"backoffInterval,omitempty"`
+	BackOffLimit        time.Duration   `json:"backoffLimit,omitempty"`
+}
+
+type openSearchLane struct {
+	logLane
+	openSearchConnection *openSearchConnection
+	metadata             map[string]string
+}
+
+type openSearchConnection struct {
+	client             *opensearchapi.Client
+	mu                 sync.Mutex
+	logBuffer          []*OpenSearchLogMessage
+	stopCh             chan *sync.WaitGroup
+	wakeCh             chan bool
+	refCount           int
+	config             *OslConfig
+	emergencyFn        OslEmergencyFn
+	ctx                context.Context
+	cancelFn           context.CancelFunc
+	messagesQueued     int
+	messagesSent       int
+	messagesSentFailed int
+}
+
+type OpenSearchLane interface {
+	Lane
+	LaneMetadata
+	ReConnect(config *OslConfig) (err error)
+	SetEmergencyHandler(emergencyFn OslEmergencyFn)
+	Stats() (stats OslStats)
+}
+
+func NewOpenSearchLane(ctx context.Context, config *OslConfig) (l OpenSearchLane) {
+	ll := deriveLogLane(nil, ctx, []Lane{}, "")
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	osl := openSearchLane{
+		openSearchConnection: &openSearchConnection{
+			logBuffer: make([]*OpenSearchLogMessage, 0),
+			stopCh:    make(chan *sync.WaitGroup, 1),
+			wakeCh:    make(chan bool, 1),
+			ctx:       ctx,
+			cancelFn:  cancelFn,
+		},
+	}
+	osl.openSearchConnection.attach()
+
+	ll.setFlagsMask(log.Ldate | log.Ltime)
+	ll.clone(&osl.logLane)
+	osl.logLane.writer = log.New(&osl, "", 0)
+
+	val := reflect.ValueOf(config)
+	if val.Kind() != reflect.Struct {
+		config = &OslConfig{
+			offline: true,
+		}
+	}
+
+	err := osl.connect(config)
+	if err != nil {
+		osl.Error(err)
+	}
+
+	go osl.openSearchConnection.processConnection()
+
+	l = &osl
+
+	return
+
+}
+
+func (osl *openSearchLane) connect(config *OslConfig) (err error) {
+	var client *opensearchapi.Client
+
+	if !config.offline && osl.openSearchConnection.client != nil && osl.openSearchConnection.client.Client != nil {
+		osl.openSearchConnection.wakeCh <- true
+	}
+
+	if !config.offline {
+		client, err = newOpenSearchClient(config.OpenSearchUrl, config.OpenSearchPort, config.OpenSearchUser, config.OpenSearchPass, config.OpenSearchTransport)
+		if err != nil {
+			return
+		}
+	}
+	osl.openSearchConnection.client = client
+	osl.openSearchConnection.config = config
```

Comment:
Bug - line 132 queues a change to `osl.openSearchConnection.config`, and it might not have run yet because the completion is not synchronized here